staking.md

```diff
@@ -0,0 +1,606 @@
+# Staking
+
+
+#### **Constants:**
+
+- **`MINIMUM_STAKE`**: The minimum amount of tokens required to be eligible for leader selection in any given epoch.
+
+#### **Participant's Balances:**
+
+- **Total Balance (TB)**: The total number of tokens a participant has deposited in the staking contract via the `stake()` function.
+- **Locked Balance (LB)**: The max of staked amounts over the **current staking cycle**, i.e., epochs `N` to `N + C`, where `C` is the number of epochs in the staking cycle (in our case, `C = 3`).
+- **Available Balance (AB)**: The portion of the Total Balance that is not locked in the current staking cycle.
+  - **AB = TB - LB**
+- **Staking Schedule (SC)**: A mapping of epochs to the amount of tokens the participant has committed to stake in each of those epochs and onwards if not changed. As an orderd list ({epoch, stake amount, locked amount})
+
+Automatic Restaking: Unless adjusted, the stake for an epoch automatically continues into future cycles.
+
+---
+
+#### **API Functions**
+
+1. **`deposit(producer:pubkey)`** (Payable Endpoint)
+   - **Description**: Allows a participant to deposit tokens into the staking contract, increasing their Total Balance (TB) and Available Balance (AB).
+   - **Parameters**:
+     - producer - the pubkey of the producer who can become leader and then produce and sign blocks.
+     -  (the amount is determined by the value of tokens sent with the transaction).
+   - **When Callable**: At any time.
+   - **Behavior**:
+     - Increases TB by the amount of tokens sent.
+     - Increases AB by the same amount.
+
+2. **`withdraw(producer:pubkey, amount:unsigned integer)`**
+   - **Description**: Allows a participant to withdraw tokens from their Available Balance (AB).
+   - **Parameters**:
+     - `producer`: The pubkey of the producer, must match the signature of the call transaction (do we need it).
+     - `amount`: The amount of tokens to withdraw.
+   - **When Callable**: At any time.
+   - **Constraints**:
+     - `amount` must be less than or equal to AB.
+     - Cannot withdraw tokens locked in the current staking cycle (i.e., tokens committed in SS for epochs `N` to `N + C`).
+
+3. **`adjustStake(producer:pubkey, amount:integer)`**
+   - **Description**: Adjusts the amount of tokens a participant wishes to stake for the **next staking cycle**.
+   - **Parameters**:
+     - `producer`: The pubkey of the producer, must match the signature of the call transaction (do we need it).
+     - `amount`: The amount to adjust (positive to increase stake, negative to decrease stake).
+   - **When Callable**: Only during the **Staking Epoch** of a cycle, but every epoch is the staking epoch of some cycle, so syou can always call it.
+   - **Constraints**:
+     - Adjustments made in the current epoch `N` affect the staking cycle starting in epoch `N`, with leader election in epoch  `N + 1`.
```

Comment:
I'm not sure what `N+1` means here, if you alter your stake that will earliest affect the leader schedule three or four epochs down the line?

Reply:
Fixed I think.